### PR TITLE
Writing Flow: Consider horizontal handled by stopPropagation in RichText

### LINKED ISSUE #5095: Inline boundary and arrow navigation edge case
Tested on current master.

In the current demo content, there's a paragraph that contains italic text:

screen shot 2018-02-16 at 09 29 04

TinyMCE adds "boundaries" to the italic text so to actually exit the boundary, you have to move the caret using the right arrow key. At that point, arrows navigation through blocks kicks in and the caret moves to the following block.

There's no way to exit the boundary not even using the mouse. Pressing the spacebar doesn't work, as you're still inside the boundary.

The only way I've found to exit the boundary and start typing not italic text is to press the right arrow (this brings me in the following block) then press the left arrow (this brings me back to the paragraph block, outside of the boundary).

Wondering what the best option would be in this case. I'd agree it's an edge case but I guess it won't be so uncommon that a block ends with some link, italic, or bold text.

### PR DESCRIPTION
**Pull request by aduth**
Extracted from #6467
Fixes #5095

This pull request seeks to stop propagation of horizontal arrow key down events when it's inferred that this is to be handled by TinyMCE in traversing out of an inline boundary. This avoids an issue where TinyMCE moves the caret, then the WritingFlow proceeds to consider the arrow navigation, leading to it being impossible to arrow-navigate out of an inline boundary at the end of a paragraph (since TinyMCE moves the caret out, then WritingFlow moves the caret to the next block).

__Testing instructions:__

Repeat steps to reproduce from #5095, noting that you can escape out of the italicized text by pressing ArrowRight at the end of the paragraph.

### CODE REVIEW

**Comment by tofumatt** [comment]
I'm taking a look at this and fixing the merge conflict, so I'll be pushing a rebased branch in a bit.

File: editor/components/rich-text/index.js [hunk 1]
```diff
@@ -433,6 +448,42 @@ export class RichText extends Component {
...
+		const isReverse = event.keyCode === LEFT;
+
+		let offset = focusOffset;
+		if ( isReverse ) {
+			offset--;
+		}
+
+		// [WORKAROUND]: When in a new paragraph in a new inline boundary node,
+		// while typing the zero-width space occurs as the first child instead
+		// of at the end of the inline boundary where the caret is. This should
```
Comment (tofumatt) [inline editor/components/rich-text/index.js]: I don't quite follow the sentence "This should only be exempt when focusNode is not _only_ the ZWSP, which occurs when caret is placed on the right outside edge of inline boundary."

Could you rephrase it to explain what this is saying? I don't quite follow.

**Review by tofumatt — APPROVED** [review]
I tested this locally and it works great. I updated the code to use the `keycodes` package and fixed the merge conflict. If you're happy with my changes I say :shipit:

File: editor/components/rich-text/index.js (hunk 1 above)
Comment (aduth) [reply editor/components/rich-text/index.js]: Heh, I was reading this again earlier and was having a hard time following my own explanation, so yes, I definitely agree it could do for some rephrasing.

**Comment by aduth** [comment]
Thanks for the review and updates @tofumatt . I'm not sure I agree with moving the ZWSP constant to the keycodes module, for two reasons:

- It isn't a [key code](https://www.w3.org/TR/uievents/#legacy-key-models), it's a character
- The `keycodes` module should be generic and not have any built-in awareness of dependencies like TinyMCE

That said, this sadly [isn't the only occasion](https://github.com/WordPress/gutenberg/search?q=ufeff&unscoped_q=ufeff) where we reference the value, so I can certainly appreciate the desire to have a single source of truth for it. Internally, TinyMCE does [expose some utilities](https://github.com/tinymce/tinymce/blob/master/src/core/main/ts/text/Zwsp.ts), though in scouring their public API I don't think it is made available for us to consume.

Alas, my intuition is to move this back into `RichText`, though in general I am very motivated to try to consolidate / simplify these "cleaning" behaviors (see also #6467, #7583).

Thoughts?

In the meantime, I also want to add a simple end-to-end test here. I should be able to salvage something from #6467.

**Comment by tofumatt** [comment]
Oh, fair enough on it not really being a keycde. `RichText` works for me.

File: editor/components/rich-text/index.js (hunk 1 above)
Comment (aduth) [reply editor/components/rich-text/index.js]: Aside: Created issue for this in TinyMCE at https://github.com/tinymce/tinymce/issues/4472

**Comment by aduth** [comment]
Updates:

- Moved constant back to `RichText` in a79442b
- Clarified the workaround comment in 48da235
- Added an end-to-end test for inline boundary traversal in 81a1b7d

My only remaining hesitation is a reluctance against stopping propagation. For example, this could inadvertently prevent [`ObserveTyping`](https://github.com/WordPress/gutenberg/tree/master/editor/components/observe-typing) from doing its thing. A few alternatives I explored (unsuccessfully):

- Adding a property to the event object to flag it as specifically handling a horizontal navigation, which is then inspected by `WritingFlow`. This requires a shared knowledge of the property between both components, and further doesn't work, since there's an incompatibility between TinyMCE events / DOM events / React events where the assigned property is lost by the time it reaches `WritingFlow`.
- Letting TinyMCE's horizontal navigation take effect before determining whether caret has moved. This is more promising, though based on the order of events firing, might not be possible for us in KeyDown and would require binding to an event like KeyUp instead, where there'd be a noticeable and undesirable (unnatural) delay for the user in traversing from one paragraph to the next.

**Comment by tofumatt** [comment]
From what you've written here, the second option sounds less complex **and** easier to implement. I'd be curious to see how "unnatural" the delay in changing paragraphs was–maybe it's not so bad? Or maybe it's quite bad and not worth pursuing. 🤷‍♂️

**Comment by aduth** [comment]
I was so very close to an amazingly simple solution, which would be to rely on TinyMCE preventing the default on the event when it's handled internally, which could be inspected from `WritingFlow` to determine whether a transition should be aborted early.

https://github.com/tinymce/tinymce/blob/0f7a0f12667bde6eae9377b50b797f4479aa1ac7/src/core/main/ts/keyboard/ArrowKeys.ts#L38

However, this seems to not work in all cases, namely at the outside edge of an inline boundary, where it falsely reports as handling an ArrowRight. Reported issue at https://github.com/tinymce/tinymce/issues/4476.

**Comment by aduth** [comment]
In retrospect though, I like the idea of using `preventDefault` as the indicator of whether WritingFlow should abort its normal flow, rather than being so forceful with `stopPropagation` in allowing it to see the event at all. It'll be challenging to get it to work this way with TinyMCE's false reporting though. Continuing to investigate...

**Comment by aduth** [comment]
Ultimately I landed at a solution which is like the first idea in https://github.com/WordPress/gutenberg/pull/6712#issuecomment-401348841 , and with full expectation that eventually we could drop all of this and rely on `Event#defaultPrevented`. The solution required binding to the native DOM document, rather than through TinyMCE's dispatched event, so that the property could be assigned.

**Comment by aduth** [comment]
Oof, I forgot to account for the fact that `this.editor.dom.doc` in the context of our inline editor is the entire page document, not just the scope of the editor node. I'll have to fix this up, as I expect it could lead to either unintended behaviors or at least wasted effort in monitoring events. It also needs to be unbound. 🤦‍♂️
